src/python_testing/TC_IDM_2_2.py

```diff
@@ -0,0 +1,617 @@
+        asserts.assert_in(descriptor_obj, returned_endpoints, "Descriptor cluster not in output")
+
+        # Step 3
+
+        # TH sends the Read Request Message to the DUT to read an attribute from a cluster at all Endpoints
+        # AttributePath = [[Cluster = Specific ClusterID, Attribute = Specific Attribute]]
+        # On receipt of this message, DUT should send a report data action with the attribute value from all the Endpoints to the DUT.
+        self.print_step(3, "Send Request Message to read one attribute on a given cluster at all endpoints")
+        # endpoint = [Clusters.Objects.Descriptor.Attributes.ServerList]
+        read_request_3 = await self.default_controller.ReadAttribute(self.dut_node_id, [server_list_attr])
+        for i in range(3):
+            returned_endpoints = read_request_3[i].keys()
+
+            # returned_endpoints = dict_keys([<class 'chip.clusters.Objects.Descriptor'>])
+            # Check if chip.clusters.Objects.Descriptor is in output
+            asserts.assert_in(descriptor_obj, returned_endpoints, "Descriptor cluster not in output")
+            # Check if ServerList is in nested output
+            asserts.assert_in(server_list_attr, read_request_3[i][descriptor_obj], "ServerList not in output")
+
+        # Step 4
+
+        # TH sends the Read Request Message to the DUT to read a global attribute from all clusters at that Endpoint
+        # AttributePath = [[Endpoint = Specific Endpoint, Attribute = Specific Global Attribute]]
+        # On receipt of this message, DUT should send a report data action with the attribute value from all the clusters to the DUT.
+        self.print_step(4, "Send Request Message to read one global attribute from all clusters at that endpoint")
+
+        # endpoint = [0, Clusters.Objects.Descriptor.Attributes.AttributeList] # Is this a global attribute? Trial and error, but looks like it
+        read_request_4 = await self.default_controller.ReadAttribute(self.dut_node_id, attribute_list_path)
+        for i in range(3):
+            returned_endpoints = read_request_4[i].keys()
+            # Check if chip.clusters.Objects.Descriptor is in output
+            asserts.assert_in(descriptor_obj, returned_endpoints, "Descriptor cluster not in output")
+            # Check if AttributeList is in nested output
+            asserts.assert_in(attribute_list, read_request_4[i][descriptor_obj], "AttributeList not in output")
+
+        # Step 5
+        # TH sends the Read Request Message to the DUT to read all attributes from all clusters on all Endpoints
+        ### AttributePath = [[]]
+        # On receipt of this message, DUT should send a report data action with the attribute value from all the clusters to the DUT.
+        self.print_step(5, "Send Request Message to read all attributes from all clusters on all endpoints")
+        read_request_5 = await self.default_controller.ReadAttribute(self.dut_node_id, [()])
+
+        for i in range(3):
```

Comment:
This check is actually covered already in one of the other IDM tests, so just check expected clusters, expected endpoints, and check that at least one cluster has > 1 attribute then put a note that this is checked in its entirety in IDM-10.1. It's actually a fairly complex check because of MEI attributes.